chassis_private.go

```diff
@@ -0,0 +1,140 @@
+ * limitations under the License.
+ **/
+
+package goovn
+
+import (
+	"fmt"
+	"github.com/ebay/libovsdb"
+)
+
+// Chassis_Private table OVN SB
+type ChassisPrivate struct {
+	UUID       string
+	ExternalID map[interface{}]interface{}
+	Name       string
+	NbCfg      int
+}
+
+func (odbi *ovndb) chassisPrivateAddImp(chName string,
+	external_ids map[string]string) (*OvnCommand, error) {
+
+	if len(chName) == 0 {
+		return nil, fmt.Errorf("chassis name cannot be empty")
+	}
+
+	namedUUID, err := newRowUUID()
+	if err != nil {
+		return nil, err
+	}
+	// row to insert
+	chPrivate := make(OVNRow)
+	chPrivate["name"] = chName
+
+	if uuid := odbi.getRowUUID(TableChassisPrivate, chPrivate); len(uuid) > 0 {
+		return nil, ErrorExist
+	}
+
+	if external_ids != nil {
+		oMap, err := libovsdb.NewOvsMap(external_ids)
+		if err != nil {
+			return nil, err
+		}
+		chPrivate["external_ids"] = oMap
+	}
+	insertOp := libovsdb.Operation{
+		Op:       opInsert,
+		Table:    TableChassisPrivate,
+		Row:      chPrivate,
+		UUIDName: namedUUID,
+	}
+	operations := []libovsdb.Operation{insertOp}
+	return &OvnCommand{operations, odbi, make([][]map[string]interface{}, len(operations))}, nil
+}
+
+func (odbi *ovndb) chassisPrivateDelImp(name string) (*OvnCommand, error) {
+	var operations []libovsdb.Operation
+
+	condition := libovsdb.NewCondition("name", "==", name)
+	deleteOp := libovsdb.Operation{
+		Op:    opDelete,
+		Table: TableChassisPrivate,
+		Where: []interface{}{condition},
+	}
+	operations = append(operations, deleteOp)
+	return &OvnCommand{operations, odbi, make([][]map[string]interface{}, len(operations))}, nil
+}
+
+func (odbi *ovndb) chassisPrivateListImp() ([]*ChassisPrivate, error) {
+	odbi.cachemutex.RLock()
+	defer odbi.cachemutex.RUnlock()
+
+	cacheChassisPrivate, ok := odbi.cache[TableChassisPrivate]
+
+	if !ok {
+		return nil, ErrorSchema
+	}
+
+	listChassisPrivate := make([]*ChassisPrivate, 0, len(cacheChassisPrivate))
+	for uuid := range cacheChassisPrivate {
+		chPrivate, err := odbi.rowToChassisPrivate(uuid)
+		if err != nil {
+			return nil, err
+		}
+		listChassisPrivate = append(listChassisPrivate, chPrivate)
+	}
+	return listChassisPrivate, nil
+}
+
+func (odbi *ovndb) chassisPrivateGetImp(chassis string) ([]*ChassisPrivate, error) {
+	var listChassisPrivate []*ChassisPrivate
+
+	odbi.cachemutex.RLock()
+	defer odbi.cachemutex.RUnlock()
+
+	cacheChassisPrivate, ok := odbi.cache[TableChassisPrivate]
+
+	if !ok {
+		return nil, ErrorSchema
+	}
+
+	for uuid, drows := range cacheChassisPrivate {
+		if chName, ok := drows.Fields["name"].(string); ok && chName == chassis {
+			chPrivate, err := odbi.rowToChassisPrivate(uuid)
+			if err != nil {
+				return nil, err
+			}
+			listChassisPrivate = append(listChassisPrivate, chPrivate)
+		}
+	}
+	return listChassisPrivate, nil
+}
+
+func (odbi *ovndb) rowToChassisPrivate(uuid string) (*ChassisPrivate, error) {
+
+	cacheChassisPrivate, ok := odbi.cache[TableChassisPrivate][uuid]
```

Comment:
odbi.cachemutex.RLock()
...
odbi.cachemutex.RUnlock()

Reply:
@vtolstov **rowToChassisPrivate** is being called from **ChassisPrivateGet** & **ChassisPrivateGet**
adding  odbi.cachemutex.RLock in **rowToChassisPrivate** will lead to scenario right as we already have locked **rowToChassisPrivate**?